File: src/misaxx-core/src/misaxx/core/filesystem/misa_filesystem_entry.cpp
/**
 * Copyright by Ruman Gerst
 * Research Group Applied Systems Biology - Head: Prof. Dr. Marc Thilo Figge
 * https://www.leibniz-hki.de/en/applied-systems-biology.html
 * HKI-Center for Systems Biology of Infection
 * Leibniz Institute for Natural Product Research and Infection Biology - Hans Knöll Insitute (HKI)
 * Adolf-Reichwein-Straße 23, 07745 Jena, Germany
 *
 * This code is licensed under BSD 2-Clause
 * See the LICENSE file provided with this code for the full license.
 */

#include <misaxx/core/filesystem/misa_filesystem_entry.h>
#include <misaxx/core/utils/filesystem.h>

using namespace misaxx;

misa_filesystem_entry::misa_filesystem_entry(std::string t_name, misa_filesystem_entry_type t_type,
                                             misa_filesystem_entry::path t_custom_external) :
        name(std::move(t_name)), type(t_type), custom_external(std::move(t_custom_external)) {
}

misa_filesystem_entry::path misa_filesystem_entry::internal_path() const {
    if (parent.expired()) {
        return name;
    } else {
        return parent.lock()->internal_path() / name;
    }
}

bool misa_filesystem_entry::has_external_path() const {
    return !external_path().empty();
}

misa_filesystem_entry::path misa_filesystem_entry::external_path() const {
    if (!custom_external.empty())
        return custom_external;
    else if (!parent.expired())
        return parent.lock()->external_path() / name;
    else
        throw std::runtime_error("Path " + internal_path().string() + " has no external path!");
}

filesystem::entry misa_filesystem_entry::self() {
    return shared_from_this();
}

filesystem::const_entry misa_filesystem_entry::self() const {
    return shared_from_this();
}

filesystem::entry misa_filesystem_entry::create(std::string t_name, misa_filesystem_entry::path t_custom_external) {
    return insert(std::make_shared<misa_filesystem_entry>(std::move(t_name), type, std::move(t_custom_external)));
}

filesystem::entry misa_filesystem_entry::insert(filesystem::entry ptr) {
    ptr->parent = self();
    children.insert({ptr->name, ptr});
    return ptr;
}

misa_filesystem_entry::iterator misa_filesystem_entry::begin() {
    return children.begin();
}

misa_filesystem_entry::const_iterator misa_filesystem_entry::begin() const {
    return children.begin();
}

misa_filesystem_entry::iterator misa_filesystem_entry::end() {
    return children.end();
}

misa_filesystem_entry::const_iterator misa_filesystem_entry::end() const {
    return children.end();
}

misa_filesystem_entry::iterator misa_filesystem_entry::find(const std::string &t_name) {
    return children.find(t_name);
}

misa_filesystem_entry::const_iterator misa_filesystem_entry::find(const std::string &t_name) const {
    return children.find(t_name);
}

bool misa_filesystem_entry::empty() const {
    return children.empty();
}

void misa_filesystem_entry::ensure_external_path_exists() const {
    if (!has_external_path())
        throw std::runtime_error("This VFS folder has no external path!");
    if (!boost::filesystem::exists(external_path()))
        boost::filesystem::create_directories(external_path());
}

filesystem::entry misa_filesystem_entry::resolve(boost::filesystem::path t_segment) {
    t_segment.remove_trailing_separator();
    if (t_segment.empty())
        return self();

    misa_filesystem_entry *current = this;

    // Navigate to subfolders if needed
    for (const auto &seg : t_segment.parent_path()) {
        auto it = current->find(seg.string());
        if (it == end()) {
            current = create(seg.string()).get();
        } else {
            current = it->second.get();
        }
    }

    // Create / access the target element
    auto it = current->find(t_segment.filename().string());
    if (it == end()) {
        return create(t_segment.filename().string());
    } else {
        return it->second;
    }
}

filesystem::const_entry misa_filesystem_entry::at(boost::filesystem::path t_segment) const {
    t_segment.remove_trailing_separator();
    if (t_segment.empty())
        return self();

    const misa_filesystem_entry *current = this;
    auto current_segment_it = t_segment.begin();

    // Navigate to subfolders if needed
    for (const auto &seg : t_segment.parent_path()) {
        auto it = current->find(seg.string());
        if (it == end()) {
            throw std::runtime_error("Cannot access path " + (internal_path() / t_segment).string());
        } else {
            current = it->second.get();
        }
    }

    // Create / access the target element
    auto it = current->find(t_segment.filename().string());
    if (it == end()) {
        throw std::runtime_error("Cannot access path " + (internal_path() / t_segment).string());
    } else {
        return it->second;
    }
}

bool misa_filesystem_entry::has_subpath(boost::filesystem::path t_segment) const {
    t_segment.remove_trailing_separator();
    if (t_segment.empty())
        return true;

    const misa_filesystem_entry *current = this;
    auto current_segment_it = t_segment.begin();

    // Navigate to subfolders if needed
    for (const auto &seg : t_segment.parent_path()) {
        auto it = current->find(seg.string());
        if (it == end()) {
            return false;
        } else {
            current = it->second.get();
        }
    }

    // Create / access the target element
    auto it = current->find(t_segment.filename().string());
    return !(it == end());
}

void misa_filesystem_entry::from_json(const nlohmann::json &) {
    throw std::runtime_error("Not implemented");
}

void misa_filesystem_entry::to_json(nlohmann::json &t_json) const {
    misa_serializable::to_json(t_json);
    throw std::runtime_error("Not implemented");
}

void misa_filesystem_entry::to_json_schema(misa_json_schema_property &t_schema) const {
    misa_serializable::to_json_schema(t_schema);
    t_schema["external-path"].declare<std::string>().make_optional("");
    for(const auto &kv : children) {
        kv.second->to_json_schema(t_schema["children"][kv.first]);
    }
    metadata->to_json_schema(t_schema["metadata"]);
}

void misa_filesystem_entry::build_serialization_id_hierarchy(std::vector<misa_serialization_id> &result) const {
    misa_serializable::build_serialization_id_hierarchy(result);
    result.emplace_back(misa_serialization_id("misa", "filesystem/entry"));
}

std::shared_ptr<misa_filesystem_entry>
misa_filesystem_entry::find_external_path(const boost::filesystem::path &t_path) {
    if(t_path == external_path()) {
        return self();
    }
    else {
        auto relative = misaxx::utils::relativize_to_direct_parent(external_path(), t_path);
        size_t d = get_depth();
        std::shared_ptr<misa_filesystem_entry> result;

        if(!relative.empty()) {
            result = self();
        }

        // Look if a child path finds a better alternative
        // Try to find the path with the highest depth
        for(const auto &kv : children) {
            auto by_child = kv.second->find_external_path(t_path);
            if(static_cast<bool>(by_child)) {
                size_t d2 = by_child->get_depth();
                if(d2 > d) {
                    result = by_child;
                    d = d2;
                }
            }
        }

        return result;
    }
}

size_t misa_filesystem_entry::get_depth() const {
    auto ptr = parent.lock();
    if(static_cast<bool>(ptr)) {
        return ptr->get_depth() + 1;
    }
    else {
        return 0;
    }
}

std::vector<filesystem::entry> misa_filesystem_entry::traverse() {
    std::vector<filesystem::entry> result;
    std::stack<filesystem::entry> stack;
    stack.push(self());

    while(!stack.empty()) {
        auto top = stack.top();
        stack.pop();
        result.push_back(top);

        for(const auto &kv : *top) {
            stack.push(kv.second);
        }
    }

    return result;
}

bool misa_filesystem_entry::remove(const std::string &t_name) {
    auto it = children.find(t_name);
    if(it != children.end()) {
        children.erase(it);
        return true;
    }
    return false;
}

File: src/misa-imagej/src/main/java/org/hkijena/misa_imagej/api/MISAFilesystemEntry.java
/*
 * Copyright by Ruman Gerst
 * Research Group Applied Systems Biology - Head: Prof. Dr. Marc Thilo Figge
 * https://www.leibniz-hki.de/en/applied-systems-biology.html
 * HKI-Center for Systems Biology of Infection
 * Leibniz Institute for Natural Product Research and Infection Biology - Hans Knöll Insitute (HKI)
 * Adolf-Reichwein-Straße 23, 07745 Jena, Germany
 *
 * This code is licensed under BSD 2-Clause
 * See the LICENSE file provided with this code for the full license.
 */

package org.hkijena.misa_imagej.api;

import com.google.gson.JsonObject;
import org.hkijena.misa_imagej.MISAImageJRegistryService;
import org.hkijena.misa_imagej.api.json.JSONSchemaObject;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Stack;

/**
 * A MISA++ filesystem entry
 */
public class MISAFilesystemEntry implements Cloneable {
    private String name;
    private Map<String, MISAFilesystemEntry> children = new HashMap<>();
    private MISAFilesystemEntry parent;
    private Path externalPath = null;
    private JSONSchemaObject metadata;
    private MISACacheIOType ioType;

    public MISAFilesystemEntry(MISAFilesystemEntry parent, String name, MISACacheIOType ioType) {
        this.parent = parent;
        this.name = name;
        this.ioType = ioType;
    }

    public MISAFilesystemEntry(MISAFilesystemEntry parent, JSONSchemaObject sourceObject, String name , MISACacheIOType ioType) {
        this(parent, name, ioType);

        if(sourceObject.hasPropertyFromPath("external-path")) {
            String pathString = sourceObject.getPropertyFromPath("external-path").getDefaultValue().toString();
            if(pathString.isEmpty()) {
                externalPath = null;
            }
            else {
                externalPath = Paths.get(pathString);
            }
        }
        if(sourceObject.hasPropertyFromPath("metadata")) {
            setMetadata((JSONSchemaObject) sourceObject.getPropertyFromPath("metadata").clone());
        }
        if(sourceObject.hasPropertyFromPath("children")) {
            for(Map.Entry<String, JSONSchemaObject> entry : sourceObject.getPropertyFromPath("children").getProperties().entrySet()) {
                children.put(entry.getKey(), new MISAFilesystemEntry(this, entry.getValue(), entry.getKey(), ioType));
            }
        }
    }

    /**
     * Clones this filesystem entry and its children.
     * Please note that the parent will be set to null. Children will be cloned as well and have their parent field set to the
     * result.
     * @return
     */
    @Override
    public Object clone() {
        MISAFilesystemEntry entry = new MISAFilesystemEntry(parent, name, ioType);
        entry.name = name;
        entry.externalPath = externalPath;
        entry.ioType = ioType;
        entry.parent = null;
        if(getMetadata() != null)
            entry.setMetadata((JSONSchemaObject) getMetadata().clone());
        for(Map.Entry<String, MISAFilesystemEntry> child : children.entrySet()) {
            MISAFilesystemEntry copy = (MISAFilesystemEntry)child.getValue().clone();
            copy.parent = entry;
            entry.children.put(child.getKey(), copy);
        }
        return entry;
    }

    public void findCaches(MISASample sample, List<MISACache> result) {
        MISACache cache = MISAImageJRegistryService.getInstance().getCacheRegistry().getCacheFor(sample,this);
        if(cache.isValid())
            result.add(cache);
        for(MISAFilesystemEntry entry : children.values()) {
            entry.findCaches(sample, result);
        }
    }

    public Path getInternalPath() {
        if(parent != null)
            return parent.getInternalPath().resolve(name);
        else
            return Paths.get(name);
    }

    public Path getExternalPath() {
        if(externalPath != null) {
            return externalPath;
        }
        else if(parent != null) {
            return parent.getExternalPath().resolve(name);
        }
        else {
            return null;
        }
    }

    public void setExternalPath(Path path) {
        externalPath = path;
    }

    /**
     * Sets the external path from a parameter JSON object that corresponds to this entry.
     * Will recursively descend
     * @param json
     */
    public void setExternalPathFromJson(JsonObject json) {
        if(json.has("external-path")) {
            externalPath = Paths.get(json.get("external-path").getAsString());
        }
        for(Map.Entry<String, MISAFilesystemEntry> kv : children.entrySet()) {
            if(json.has("children") && json.getAsJsonObject("children").has(kv.getKey())) {
                kv.getValue().setExternalPathFromJson(json.getAsJsonObject("children").getAsJsonObject(kv.getKey()));
            }
        }
    }

    /**
     * Returns true if there are external path definitions
     * @return
     */
    public boolean hasExternalPathDefinitions() {
        Stack<MISAFilesystemEntry> stack = new Stack<>();
        stack.push(this);

        while(!stack.isEmpty()) {
            MISAFilesystemEntry top = stack.pop();
            if(top.externalPath != null)
                return true;
            for(MISAFilesystemEntry entry : top.children.values()) {
                stack.push(entry);
            }
        }
        return false;
    }

    /**
     * Converts this filesystem enty into a proper MISA++ JSON entry
     * @return
     */
    public JsonObject toJson() {
        JsonObject object = new JsonObject();
        if(externalPath != null) {
            object.addProperty("external-path", getExternalPath().toString());
        }
        if(getMetadata() != null && getMetadata().wasChanged()) {
            object.add("metadata", getMetadata().toJson());
        }
        if(children != null) {
            JsonObject childrenJson = new JsonObject();
            for(Map.Entry<String, MISAFilesystemEntry> kv : children.entrySet()) {
                childrenJson.add(kv.getKey(), kv.getValue().toJson());
            }
            object.add("children", childrenJson);
        }
        return object;
    }

    public MISAFilesystemEntry resolve(String ...segments) {
        MISAFilesystemEntry current = this;
        for(String segment : segments) {
            if(!current.children.containsKey(segment)) {
                MISAFilesystemEntry entry = new MISAFilesystemEntry(this, segment, ioType);
                current.children.put(segment, entry);
            }
            current = current.children.get(segment);
        }
        return current;
    }

    @Override
    public String toString() {
        return getInternalPath().toString();
    }

    public void update() {
        for(Map.Entry<String, MISAFilesystemEntry> child : children.entrySet()) {
            child.getValue().name = child.getKey();
            child.getValue().update();
        }
    }

    public JSONSchemaObject getMetadata() {
        return metadata;
    }

    public void setMetadata(JSONSchemaObject metadata) {
        this.metadata = metadata;
    }

    public MISACacheIOType getIoType() {
        return ioType;
    }

    public String getName() {
        return name;
    }

    public void setParent(MISAFilesystemEntry parent) {
        this.parent = parent;
        if(parent != null) {
            parent.update();
        }
    }

    public MISAFilesystemEntry getParent() {
        return parent;
    }

    public void insert(String name, MISAFilesystemEntry importedFilesystem) {
        this.children.put(name, importedFilesystem);
        update();
    }

    public boolean remove(String name) {
        if(this.children.containsKey(name)) {
            this.children.remove(name);
            return true;
        }
        return false;
    }
}
